### PR TITLE
ENH: stats.fit: add maximum spacing estimation

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
None

#### What does this implement/fix?
From the [Wikipedia entry](https://en.wikipedia.org/wiki/Maximum_spacing_estimation):
> There are certain distributions, especially those with three or more parameters, whose likelihoods may become infinite along certain paths in the parameter space. Using maximum likelihood to estimate these parameters often breaks down, with one parameter tending to the specific value that causes the likelihood to be infinite, rendering the other parameters inconsistent. The method of maximum spacings, however, being dependent on the difference between points on the cumulative distribution function and not individual likelihood points, does not have this issue, and will return valid results over a much wider array of distributions

This adds maximum spacing estimation as an option for `rv_continuous`. 

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(1638083107694713882823079058616272161)

params = (2, 3)
dist = stats.rayleigh
rvs = dist(*params).rvs(size=1000, random_state=rng)

res = stats.fit(dist, rvs, bounds={'loc': (-10, 10), 'scale': (0.1, 10)},
                method='mse')
res.plot()
```

image

#### Additional information
TODO:
- [x] add discrete distribution support
- [x] investigate the tests (e.g. which are slow, which are xslow, and which really fail or should be skipped otherwise).
- [x] decide on https://github.com/scipy/scipy/pull/17141#pullrequestreview-1132404838
- [x] figure out why `np.errstate` is needed in tests
- [ ] ~~use `_logcdf`?~~ I don't think it's needed in this first pass. `logcdf` is useful for resolving differences between CDF values that would not be resolvable otherwise - at the tails when the CDF is less than ~1e-300 or within ~1e-16 of 1 (and only for distributions that have an accurate `logcdf`). The probability of getting one observation that is within $p=10^{-16}$ of $1$ in a dataset with $n$ observations is $1 - (1-p)^n \approx np$. So approximately one in a billion samples each of size one billion observations will contain a single observation with such a small CDF. There would need to be at least _two_ observations in a single sample for `logcdf` to matter, and this probability is obviously smaller. So I doubt it will affect the results of fitting in practice - but it would _always_ slow it down.

### CODE REVIEW

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:1615]: The only real change is to use this to penalize `log_fitfun`, which can be `self._logpxf` or another function, instead of just `self._logpxf`. The rest is changing variable names so that they're not specific to MLE.

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:2557]:
```suggestion
              Likelihood Estimate); "MPS" (Maximum Product Spacing) and
```

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:68]:
```suggestion

failing_fits = {"MM": mm_failing_fits + mm_slow_fits, "MLE": mle_failing_fits,
```

**Comment by tupui** [comment]
With the new infra coming up, isn't `scipy.stats.fit` going to be the main fit method we want to have? If that's the case, maybe we should only add enh into this one?

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:258]: Now this accepts the name of a negative log fitness function to assess the quality of the fit. For MLE, we call the negative log likelihood function to assess the fit; for MPE, we call the negative log product spacing function.

**Comment by mdhaber** [comment]
@tupui yup, as we discussed I just wanted to do the quickest thing to play with this technique. Now, I've taken this out of `stats.rv_continuous.fit` and added it to `stats.fit`. I still need to modify this to work for discrete distributions and investigate the tests (e.g. which are slow, which are xslow, and which really fail or should be skipped otherwise).

**Review by tupui — APPROVED** [review]
Thanks for the update Matt, LGTM provided we mitigate the CI issues. 2 tests are failing due to timeout, we should remove these if possible `ncf, nct, beta`. Even for xslow 60s is not reasonable I think.

Aside from backward compatibility, what would be the reason for not making this new option the default? From the wiki description, it seems like it should be more robust. But I see a lot of skip in the tests, what is happening here?

I will leave this open for a few days in case more maintainers pass by.

**Comment by mdhaber** [comment]
There are lots of skips in the MLE test, too. I'm surprised at how few skips were needed, actually. Some are because the optimization is too slow (e.g. studentized range, Levy stable). For others, the auto-generated bounds  may be too wide. Some may be fixable - I haven't investigated yet.
Even if this is a bit more robust, I'm not sure a change in default is needed because MLE is certainly more common.

**Comment by mdhaber** [comment]
Ok, now the tests of the distributions are split up into groups - slow, xslow, skip, and the rest. There is only one distribution that completes without warning yet fails the test for the objective function value, so this may be more reliable than MLE from that perspective. However, it's typically much slower. There are only a handful of distributions that are not marked slow or otherwise.

I still need to investigate the distributions that fail for other reasons (e.g. warning or error during execution) a little more closely, but often it is more of an issue with the to distribution than this new code.

ND of course I have to generalize this

**Comment by WarrenWeckesser** [comment]
I have an unreleased implementation of MPS.  I'll touch up anything that looks too messy and put it somewhere on github this weekend, then see if there is anything in there worth including in this PR (or in a follow-up).  One issue that I don't see handled here is repeated values in the input data.

**Comment by mdhaber** [comment]
Typically it doesn't but this seems like a reasonable generalization:
[Maximum Product of Spacings Method - A Unified Formulation...](
https://projecteuclid.org/journals/illinois-journal-of-mathematics/volume-43/issue-3/Maximum-product-of-spacings-method--A-unified-formulation-with/10.1215/ijm/1255985105.pdf)

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:1655]: From Wikipedia:
> For convenience also denote x(0) = −∞ and x(n+1) = +∞.

The CDF at -∞ is 0 and at +∞ it's 1, hence the concatenation.

But Shao and Hahn's generalization (for repeated observations) only adds x(0) = −∞.

image

I'm conflicted.

Comment (mdhaber) [reply scipy/stats/_distn_infrastructure.py]: The "spacing" in Maximum Spacing Estimation refers to the spacing between PDFs, that is, the difference between CDFs of points in x. For continuous distributions, we expect there to be some non-zero spacing between the CDFs of x(n) and x(n+1) = +∞, so it makes sense to add x(n+1) = +∞. For discrete distributions, though, it is not uncommon to get samples at the upper end of the support of a discrete distribution, at which the CDF is 1. But if we append x(n+1) = +∞ in this case, we get a "spacing" between the CDFs of x(n) and x(n+1) = +∞ that equals zero. This would mean that the product spacing of the true parameters is often 0 for discrete distribution. (Another way of looking at it: the negative log of this is the objective function to minimize, and it would be infinite for the true parameters.) We don't want that. So for discrete distributions, it is not appropriate to append x(n+1) = +∞.

 (When we take the log of these 0s that occur for the true parameters, we get the `invalid` `RuntimeWarning` reported [below](https://github.com/scipy/scipy/pull/17141#pullrequestreview-1132409801).)

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [reply scipy/stats/tests/test_fit.py]: `vonmises` warnings were because `vonmises` can produce a negative CDF. I don't think we need to worry about `vonmises`.

The rest were because of the issue described [above](https://github.com/scipy/scipy/pull/17141#discussion_r989710751).

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distn_infrastructure.py]: Oops I misread. The paper addresses how to handle this.

File: scipy/stats/_distn_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distn_infrastructure.py:1655]: I misread the paper before. If `cdf_data` does not contain 1, it should be appended.

**Comment by mdhaber** [comment]
Fixed the implementation of the generalization so that it works for discrete distributions and samples with repeated elements, but it also reduces (exactly) to the original method when the distribution is continuous and all observations are unique. I don't think I wanted to do anything else here (except we should probably add `rdist` back to the list of xslow tests).

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:240]:
```suggestion
                       'f', 'powerlognorm', 'ncx2', 'rdist', 'reciprocal',
```

**Comment by mdhaber** [comment]
Besides providing a good fit, here is evidence that the implementation is actually maximum spacing estimation. Compare against the [examples on Wikipedia](https://en.wikipedia.org/wiki/Maximum_spacing_estimation#Examples).

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(9843212616816518964)
from numpy.testing import assert_allclose

# Example 1
data = [2, 4]
dist = stats.expon
bounds = {'loc': (0, 0), 'scale': (1e-8, 10)}
res_mle = stats.fit(dist, data, bounds=bounds, method='mle')
assert_allclose(res_mle.params.scale, 3, atol=1e-3)
res_mse = stats.fit(dist, data, bounds=bounds, method='mse')
assert_allclose(res_mse.params.scale, 3.915, atol=1e-3)

# Example 2
dist = stats.uniform
n = 10
data = dist(3, 6).rvs(size=n, random_state=rng)
bounds = {'loc': (0, 10), 'scale': (1e-8, 10)}
res = stats.fit(dist, data, bounds=bounds, method='mse')
# (loc=3.608118420015416, scale=5.509323262055043)

x = np.sort(data)
a = (n*x[0] - x[-1])/(n - 1)
b = (n*x[-1] - x[0])/(n - 1)
ref = a, b-a  # (3.6081133632151503, 5.509328130317254)
assert_allclose(res.params, ref, atol=1e-5)
```

**Comment by mdhaber** [comment]
OK I went ahead and added that extra test to confirm that MSE is actually doing MSE. 

@tupui I see that you already approved, and [the changes](https://github.com/scipy/scipy/pull/17141/files/d317c4f2452e93bf9b846457f9f7672f2b1e9c27..b34c705939f915737241bd2bb01d2747dfbd5e89) since then have mostly been mild - for instance, figuring out how to mark each test (slow, xslow, etc.). One change to look at is the redefinition of `log_psf` to handle discrete distributions based on reference [[1]](https://projecteuclid.org/journals/illinois-journal-of-mathematics/volume-43/issue-3/Maximum-product-of-spacings-method--A-unified-formulation-with/10.1215/ijm/1255985105.full). LMK if there's anything else!

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:652]:
```suggestion
        assert_allclose(res.params, ref, atol=1e-4)
```

File: scipy/stats/tests/test_fit.py
Comment (mdhaber) [inline scipy/stats/tests/test_fit.py:652]:
```suggestion
        assert_allclose(res.params, ref, rtol=1e-4)
```
Let's commit after the Cirrus CI issues are resolved to see if we can get CI green.

**Review by tupui — APPROVED** [review]
Update LGTM, thanks Matt.

We could update the docstring of the `optimize` param to mention the new objective, but this is a nitpick. I would merge if you're fine with it now.

**Comment by mdhaber** [comment]
Thanks @tupui. I fixed it.

**Pull request merged**
